### PR TITLE
Adds Light Replacer Refilling With Box of Lights

### PR DESCRIPTION
**Pull request by Yolopanther12**
Adds the ability to refill a light replacer using a box of lights.

A box of lights specifically must be used, broken or burnt out lights will not be used, and the amount of lights used will only be as many as it takes to fill the replacer to capacity.

:cl:
rscadd: Added the ability to use a box of lights on a light replacer to more quickly refill it to capacity.
/:cl:

### CODE REVIEW

**Comment by WJohn** [comment]
Why was this never the case :+1:

**Comment by Aranclanos** [comment]
you're looping through the box contents twice and the lines where you delete the light and add an use are copypasted

**Comment by Yolopanther12** [comment]
@Aranclanos Worst case scenario is that I loop through the entire contents twice, yeah. Though it's likely the first loop will break before that happens (though I suppose this depends on how objects are added to the box, hadn't considered that). I thought about other ways to do it while still allowing the first warning message to be, well, first. Nothing came to mind however.

As to "the lines where you delete the light and add an use are copypasted" I'm not quite sure what you mean by that.

File: code/game/objects/items/devices/lightreplacer.dm [hunk 1]
```diff
@@ -98,6 +98,33 @@
 			user << "<span class='warning'>You need a working light!</span>"
 			return
 
+	if(istype(W, /obj/item/weapon/storage/box/lights))
+		var/obj/item/weapon/storage/box/lights/B = W
+		var/tmp/replacement_status = 1 // 0-useable lights found, 1-no useable lights, 2-light replacer was full
```
Comment (RemieRichards) [inline code/game/objects/items/devices/lightreplacer.dm]: the ```/tmp``` modifier does nothing here.
it's used for vars not saved into savefiles, and you're not on a class OR saving things here.

File: code/game/objects/items/devices/lightreplacer.dm [hunk 2]
```diff
@@ -98,6 +98,33 @@
 			user << "<span class='warning'>You need a working light!</span>"
 			return
 
+	if(istype(W, /obj/item/weapon/storage/box/lights))
+		var/obj/item/weapon/storage/box/lights/B = W
+		var/tmp/replacement_status = 1 // 0-useable lights found, 1-no useable lights, 2-light replacer was full
+
+		if(src.uses == max_uses)
+			if(!B.contents.len) // empty box
+				replacement_status = 1
```
Comment (RemieRichards) [inline code/game/objects/items/devices/lightreplacer.dm]: dont use magic numbers, 
```
#define REP_STAT_NO_LIGHTS 1
#define REP_STAT_FULL 2
```

File: code/game/objects/items/devices/lightreplacer.dm [hunk 3]
```diff
@@ -98,6 +98,33 @@
 			user << "<span class='warning'>You need a working light!</span>"
 			return
 
+	if(istype(W, /obj/item/weapon/storage/box/lights))
```
Comment (RemieRichards) [inline code/game/objects/items/devices/lightreplacer.dm]: you should make this ```/obj/item/weapon/storage``` or atleast ```/box```, so you can fill it from any container/any box.

Comment (JJRcop) [inline code/game/objects/items/devices/lightreplacer.dm]: >A box of lights specifically must be used

Probably intentional, why though, @Yolopanther12?

**Comment by Supermichael777** [comment]
janitor of the year

**Comment by YotaXP** [comment]
## :+1: **YES!!**  I LOVE YOU! :+1: 

But it totally should accept any box or container.

File: code/game/objects/items/devices/lightreplacer.dm (hunk 2 above)
Comment (YotaXP) [inline code/game/objects/items/devices/lightreplacer.dm]: I don't think we should pollute the symbol space for stuff that's specific to the body of a single proc.

Comment (RemieRichards) [inline code/game/objects/items/devices/lightreplacer.dm]: you can undefine them after the proc yota.

File: code/game/objects/items/devices/lightreplacer.dm [hunk 4]
```diff
@@ -98,6 +98,33 @@
...
+					else
+						break
+
+		if(replacement_status == 1)
+			user << "<span class='warning'>The [B.name] contains no useable lights!</span>"
+			return
+		else if(replacement_status == 2)
+			user << "<span class='warning'>The [src.name] is full!</span>"
+			return
+		user << "<span class='notice'>You fill the [src.name] with lights from the [B.name]. You have [uses] lights remaining.</span>"
```
Comment (YotaXP) [inline code/game/objects/items/devices/lightreplacer.dm]: This stuff should just be in a `switch` block.

File: code/game/objects/items/devices/lightreplacer.dm (hunk 2 above)
Comment (Aranclanos) [inline code/game/objects/items/devices/lightreplacer.dm]: just use true and false if you're feeling so gay

**Comment by Aranclanos** [comment]
also remove all the comments, none of them are needed

File: code/game/objects/items/devices/lightreplacer.dm (hunk 2 above)
Comment (RemieRichards) [inline code/game/objects/items/devices/lightreplacer.dm]: >binary response to a ternary situation
Yes Aran, because that's logical.

Comment (Aranclanos) [inline code/game/objects/items/devices/lightreplacer.dm]: idk i can't read the code because ITS FILLED WITH YOUR COMMENTS

Comment (RemieRichards) [inline code/game/objects/items/devices/lightreplacer.dm]: just doing my job, unlike some people.

Comment (Aranclanos) [inline code/game/objects/items/devices/lightreplacer.dm]: thats it you and me outside

**Comment by Yolopanther12** [comment]
Working on some changes to this, please do not merge.

**Comment by YotaXP** [comment]
> also remove all the comments, none of them are needed

:-1: No, keep em.  Comments are **good** to have.  For instance, without the comments, I would have no friggin idea what `if(L.status == 0)` was for.

**Comment by Yolopanther12** [comment]
Okay, I think the latest commit has appeased all concerned parties. Forgot to comment but I think it's a lot more straightforward to see what's going on even without.

File: code/game/objects/items/devices/lightreplacer.dm [hunk 5]
```diff
@@ -98,6 +98,33 @@
...
+					found_good_light = 1
+					if(src.uses < max_uses)
+						qdel(L)
+						AddUses(1)
+						replaced_something = 1
+					else
+						break
+
+		if(!found_good_light)
+			user << "<span class='warning'>The [S.name] contains no useable lights!</span>"
```
Comment (duncathan) [inline code/game/objects/items/devices/lightreplacer.dm]: `\The [S]`

File: code/game/objects/items/devices/lightreplacer.dm [hunk 6]
```diff
@@ -98,6 +98,33 @@
...
+			return
+
+		if(!replaced_something && src.uses == max_uses)
+			user << "<span class='warning'>The [src.name] is full!</span>"
```
Comment (duncathan) [inline code/game/objects/items/devices/lightreplacer.dm]: `\The [src]`

File: code/game/objects/items/devices/lightreplacer.dm [hunk 7]
```diff
@@ -98,6 +98,33 @@
...
+			return
+
+		user << "<span class='notice'>You fill the [src.name] with lights from the [S.name]. You have [uses] lights remaining.</span>"
```
Comment (duncathan) [inline code/game/objects/items/devices/lightreplacer.dm]: `\the [src]` and `\the [S]`

Comment (Yolopanther12) [inline code/game/objects/items/devices/lightreplacer.dm]: Ah, I see what this does now.

**Comment by Fox-McCloud** [comment]
I'm really surprised the light replacer doesn't just magically generate lights like an RCD does and and require recharging in a recharger as opposed to refilling it with actual glass or bulbs.

Also, it should totally be able to recycle broken glass tubes, IMO, at a discount or something.

**Comment by MrStonedOne** [comment]
fox, you can, its called a autolathe.

File: code/game/objects/items/devices/lightreplacer.dm [hunk 8]
```diff
@@ -98,6 +98,33 @@
...
+	if(istype(W, /obj/item/weapon/storage))
+		var/obj/item/weapon/storage/S = W
+		var/found_good_light = 0
+		var/replaced_something = 0
+
+		for(var/obj/item/I in S.contents)
+			if(istype(I,/obj/item/weapon/light))
+				var/obj/item/weapon/light/L = I
+				if(L.status == LIGHT_OK)
+					found_good_light = 1
```
Comment (Razharas) [inline code/game/objects/items/devices/lightreplacer.dm]: Delete this

File: code/game/objects/items/devices/lightreplacer.dm [hunk 9]
```diff
@@ -98,6 +98,33 @@
...
+					if(src.uses < max_uses)
+						qdel(L)
+						AddUses(1)
```
Comment (Razharas) [inline code/game/objects/items/devices/lightreplacer.dm]: Replace this with "found_good_light++"

File: code/game/objects/items/devices/lightreplacer.dm [hunk 10]
```diff
@@ -98,6 +98,33 @@
...
+						replaced_something = 1
+					else
+						break
+
```
Comment (Razharas) [inline code/game/objects/items/devices/lightreplacer.dm]: Add "AddUses(found_good_light)" here

File: code/game/objects/items/devices/lightreplacer.dm [hunk 11]
```diff
@@ -98,6 +98,33 @@
...
+		if(!found_good_light)
+			user << "<span class='warning'>\The [S] contains no useable lights!</span>"
+			return
+
+		if(!replaced_something && src.uses == max_uses)
```
Comment (Razharas) [inline code/game/objects/items/devices/lightreplacer.dm]: Swap this check and the one above, check for if its full should be first

**Comment by Jordie0608** [comment]
>page didn't update new comments until after i merged
Balls

**Pull request merged**